proposals/hasCrossSiteAncestor.md

```diff
@@ -0,0 +1,122 @@
+# Proposal: Update Extensions to interact with Partitioned Cookies containing Cross-Site Ancestor Chain Bit values
+
+**Summary**
+
+Allow extensions to utilize a hasCrossSiteAncestor boolean value when interacting with partitioned cookies that include a cross-site ancestor chain bit in their partiton key.
+
+#### Background information: Description of Cross-site Ancestor Chain Bit in partitioned cookies
+
+The cross-site ancestor chain bit is a component of the cookie partition key that is set by the browser when keys are created. If the bit indicates true, it means the cookie has been set in a third-party context.
```

Comment:
(continuation of thread from https://github.com/w3c/webextensions/pull/581#discussion_r1574385184)

Questions that I have that are not answered by this section:

- (When) can top-level documents have the bit?
- (When) can unpartitioned cookies have the bit?

Top-level navigations triggered from iframes (whether `_top` or a new popup) are scenarios where these questions become relevant.

Reply:
>(When) can top-level documents have the bit?

In the chromium implementation, unpartitioned cookies do not have the bit because unpartitioned cookies do not have a `partitionKey`. 

>(When) can top-level documents have the bit?

Regardless of the source of the navigation, partitioned cookies accessible from top-level documents will always have a value of `false`, indicating no cross-site ancestor, since as the top-level they do not have any ancestor. This choice was made as a result of compatibility issues detailed in this [bug](https://g-issues.chromium.org/issues/339410000).

Reply:
Thanks for your clarification. The "Unpartitioned cookies do not have the bit" part was not obvious to me, and it would be worth calling that characteristic explicitly in the proposal document. "Unpartitioned cookies always have hasCrossSiteAncestor false."

> > (When) can top-level documents have the bit?
> 
> In the chromium implementation, unpartitioned cookies do not have the bit because unpartitioned cookies do not have a `partitionKey`. In the current implementation, passing an empty value for `topLevelSite` indicates an unpartitioned cookie. To maintain backwards compatibility, I think it makes sense to have the api ignore the value of `hasCrossSiteAncestor` when the `topLevelSite` is empty. If you're ok with this I can update the language of the proposal to reflect it.

Why do you think that we should treat `hasCrossSiteAncestor` specially when `topLevelSite` is empty? I can't think of a benefit for extensions or browsers to special case this, can you?

Reply:
>Why do you think that we should treat hasCrossSiteAncestor specially when topLevelSite is empty? I can't think of a benefit for extensions or browsers to special case this, can you?

How about adding language saying, "Unpartitioned cookies, which are indicated by a cookie not containing a `partitionKey`, the key not containing a `topLevelSite` or an empty `topLevelSite`, always have a `hasCrossSiteAncestor` value of `false`.

This addresses how the value should be handled when the cookie is unpartitioned and clarifies when a cookie is considered unpartitioned.

Reply:
The proposed language SGTM. It clarifies the behavior with unpartitioned cookies.

The situation with top-level documents is still under-specified in this document. Can you also add a note about top-level documents? Based on your reply in the other thread at https://github.com/w3c/webextensions/pull/581#discussion_r1707027673, I think that Chrome can have partitioned cookies in top-level documents.

"Top-level documents may have partitioned cookies (topLevelSite non-empty, i.e. set to the site of the top-level document), but hasCrossSiteAncestor is always false for top-level contexts by design (CHIPS spec link here)."

This is mainly to help extension developers with interpreting the behavior correctly, independent of browser-specific behavior.

In Firefox, top-level documents are unpartitioned, and therefore hasCrossSiteAncestor is false (following the reasoning earlier in this thread). Third-party frames (including frames that are same-site to the top document but with a cross-site frame in between, but excluding frames whose full ancestor chain is same-site) are partitioned automatically. Third-party frames can opt in to unpartitioned cookies with `requestStorageAccess`.

In Chrome, partitioning seems to be tied to the `Partitioned` cookie attribute that the website has pass to `Set-Cookie`, instead of forced by the browser. While Firefox does not partition cookies in top-level documents, it seems that in Chrome a top-level document can still have partitioned cookies, i.e. `topLevelSite` being non-empty, if the cookie is set with the `Partitioned` attribute.

(If my understanding is wrong, please correct me - I am spelling out my thoughts process to make it easier to challenge my misconceptions if any.)

Reply:
You have the correct understanding of the differences between the implementations.

I like the language you suggested as well and will add it.